sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala

```diff
@@ -61,7 +65,23 @@ object JdbcUtils extends Logging {
         throw new IllegalStateException(
           s"Did not find registered driver with class $driverClass")
       }
-      val connection: Connection = driver.connect(options.url, options.asConnectionProperties)
+
+      val connection: Connection = {
+        if (options.keytab != null && options.principal != null) {
+          val keytabFileName = FilenameUtils.getName(options.keytab)
+          UserGroupInformation
+            .loginUserFromKeytabAndReturnUGI(options.principal, keytabFileName)
+            .doAs(new PrivilegedAction[Connection] {
+              override def run(): Connection = {
+                driver.connect(options.url, options.asConnectionProperties)
+              }
+            })
+        }
+        else {
```

Comment:
Seems there is no option for TGT nor delegation token. This later may be useful for hadoop components though. But distributing the keytab is not the most secure way and KDC may be overloaded.

Reply:
Hi @misutoth , 

1. Distributing keytab is based on the assumption of secure HDFS. Each executor requires a keytab to do kinit, so we have to distribute the keytab across different executors. If HDFS is secured, then the keytab should be safe.

2. A user is required to upload the data to /spark/admin.keytab. This file is pre-uploaded by the user. 

3. Since SQL server does not delegation token, we cannot use delegation token. If SQL server supports delegation token in the future, we can use a more efficient way to do the kinit. Right now we need to use keytab (or password) to do the kinit and this limit comes from SQL server side, not HDFS side.

Reply:
I see an issue with `UserGroupInformation.loginUserFromKeytabAndReturnUGI(options.principal, keytabFileName)`

This would replace static field on UGI for keytab and principal. So further call to UGI for for hadoop may not work.

I think better approach would be to not to rely on UGI. instead do login using LoginContext and wrap call with Subject.doAs(subject,...)

Reply:
> I see an issue with `UserGroupInformation.loginUserFromKeytabAndReturnUGI(options.principal, keytabFileName)`
> 
> This would replace static field on UGI for keytab and principal. So further call to UGI for for hadoop may not work.
> 
> I think better approach would be to not to rely on UGI. instead do login using LoginContext and wrap call with Subject.doAs(subject,...)

Good suggestion. I will give a try and if it works I will update this PR.

Reply:
@misutoth @dasbh, the keytab is only for JDBC connection and not intended for future use by other hadoop calls, for now, since these database does not support delegation token yet. Usually customer should use their own principal in keytab file stored in secure HDFS to login to sql database, the same as the credential they use to submit the Spark application. This is really a workaround for databases that do not support delegation token.